File: python/navigation/maps/occupancy_grid.py
"""OccupancyGrid2D — ROS-style grid map providing DiscreteSpace + SamplingSpace.

world<->grid conversion lives ONLY here (map layer owns the coordinate frames).
Discrete state is a Cell (row, col); sampling state is a world Point (x, y).
"""

from __future__ import annotations

import math
from collections.abc import Callable

import numpy as np

from navigation.core.capabilities import Capability, MapBase
from navigation.core.types import Cell, Footprint, Point, Pose

_SQRT2 = math.sqrt(2.0)
# Sentinel "infinity" for the 1D distance transform below. Must stay finite (an
# actual inf would produce a 0/0 -> nan in the lower-envelope intersection formula)
# but far larger than any squared distance a real grid can produce.
_EDT_INF = 1e15


def _dt_1d(f: np.ndarray) -> np.ndarray:
    """Exact 1D squared-distance lower envelope (Felzenszwalb & Huttenlocher 2004,
    "Distance Transforms of Sampled Functions", §2). `f[q]` is 0 at a source and
    `_EDT_INF` elsewhere; returns, for every q, min_{q'} (q-q')^2 + f[q'].
    """
    n = f.shape[0]
    d = np.empty(n, dtype=np.float64)
    v = np.zeros(n, dtype=np.int64)  # q-indices of parabolas kept in the envelope
    z = np.empty(n + 1, dtype=np.float64)  # envelope boundaries between kept parabolas
    k = 0
    v[0] = 0
    z[0] = -_EDT_INF
    z[1] = _EDT_INF
    for q in range(1, n):
        while True:
            s = ((f[q] + q * q) - (f[v[k]] + v[k] * v[k])) / (2.0 * q - 2.0 * v[k])
            if s <= z[k]:
                k -= 1
            else:
                break
        k += 1
        v[k] = q
        z[k] = s
        z[k + 1] = _EDT_INF
    k = 0
    for q in range(n):
        while z[k + 1] < q:
            k += 1
        d[q] = float((q - v[k]) ** 2) + f[v[k]]
    return d


def _squared_edt(non_free: np.ndarray) -> np.ndarray:
    """2-pass exact squared Euclidean distance transform to the `non_free` mask
    (Felzenszwalb & Huttenlocher 2004): 1D transform along each column, then along
    each row of the result. Distances are in cell units (squared); the caller scales
    by resolution and takes sqrt once, so cross-language results agree bit-for-bit
    up to that single final sqrt.
    """
    h, w = non_free.shape
    f = np.where(non_free, 0.0, _EDT_INF)
    partial = np.empty_like(f)
    for col in range(w):
        partial[:, col] = _dt_1d(f[:, col])
    out = np.empty_like(f)
    for row in range(h):
        out[row, :] = _dt_1d(partial[row, :])
    return out
# 8-connected moves; diagonals cost sqrt(2), orthogonals 1.0.
_MOVES_8 = [
    (-1, 0, 1.0),
    (1, 0, 1.0),
    (0, -1, 1.0),
    (0, 1, 1.0),
    (-1, -1, _SQRT2),
    (-1, 1, _SQRT2),
    (1, -1, _SQRT2),
    (1, 1, _SQRT2),
]
_MOVES_4 = [(-1, 0, 1.0), (1, 0, 1.0), (0, -1, 1.0), (0, 1, 1.0)]


class OccupancyGrid2D(MapBase):
    def __init__(
        self,
        pixels: np.ndarray,
        resolution: float,
        origin: tuple[float, float, float],
        occupied_thresh: float = 0.65,
        free_thresh: float = 0.196,
        connectivity: int = 8,
        seed: int = 0,
    ) -> None:
        if connectivity not in (4, 8):
            raise ValueError(f"connectivity must be 4 or 8, got {connectivity}")
        self._height, self._width = int(pixels.shape[0]), int(pixels.shape[1])
        self._resolution = resolution
        self._origin_x, self._origin_y = origin[0], origin[1]
        self._connectivity = connectivity
        # occupied_thresh is accepted for ROS yaml signature compatibility but not
        # used: traversability is decided by free_thresh alone (occupied/unknown are
        # blocked), so a separate occupied cutoff would be redundant.
        # occ = 1 - p/255; free iff occ <= free_thresh; anything not clearly free
        # (occupied or unknown) is non-traversable.
        occ = 1.0 - pixels.astype(np.float64) / 255.0
        self._free = occ <= free_thresh
        self._moves = _MOVES_8 if connectivity == 8 else _MOVES_4
        self._rng = np.random.default_rng(seed)
        # Lazy EDT cache (cell-unit squared distances): computed on first
        # distance_to_nearest() call so the existing global_planning suite, which
        # never touches ObstacleQuery, pays nothing for it.
        self._edt_sq: np.ndarray | None = None

    # --- dimensions -------------------------------------------------------
    @property
    def height(self) -> int:
        return self._height

    @property
    def width(self) -> int:
        return self._width

    @property
    def resolution(self) -> float:
        return self._resolution

    def free_mask(self) -> np.ndarray:
        """Boolean [H, W] mask of traversable cells (read-only view for viz)."""
        return self._free

    # --- coordinate frames (owned here only) ------------------------------
    def cell_to_world(self, row: int, col: int) -> Point:
        x = self._origin_x + (col + 0.5) * self._resolution
        y = self._origin_y + ((self._height - 1 - row) + 0.5) * self._resolution
        return (x, y)

    def world_to_cell(self, x: float, y: float) -> Cell:
        col = int(math.floor((x - self._origin_x) / self._resolution))
        row = (self._height - 1) - int(math.floor((y - self._origin_y) / self._resolution))
        return (row, col)

    def in_bounds(self, row: int, col: int) -> bool:
        return 0 <= row < self._height and 0 <= col < self._width

    def is_free_cell(self, row: int, col: int) -> bool:
        return self.in_bounds(row, col) and bool(self._free[row, col])

    # --- capabilities -----------------------------------------------------
    def capabilities(self) -> set[Capability]:
        return {
            Capability.DISCRETE_SPACE,
            Capability.SAMPLING_SPACE,
            Capability.LINE_OF_SIGHT_SPACE,
            Capability.DYNAMIC_GRID_SPACE,
            Capability.SE2_COLLISION_SPACE,
            Capability.OBSTACLE_QUERY,
        }

    # --- DiscreteSpace[Cell] ---------------------------------------------
    def _neighbors_impl(
        self, s: Cell, is_free: Callable[[int, int], bool]
    ) -> list[tuple[Cell, float]]:
        # Single 8-move + corner-cut worker shared by neighbors() (truth predicate)
        # and passable_neighbors() (belief predicate). ``is_free`` decides both cell
        # entry and the diagonal corner rule so both callers forbid corner-cutting
        # identically.
        row, col = s
        out: list[tuple[Cell, float]] = []
        for dr, dc, cost in self._moves:
            nr, nc = row + dr, col + dc
            if not is_free(nr, nc):
                continue
            # Forbid corner-cutting: a diagonal needs both shared orthogonals free.
            if dr != 0 and dc != 0:
                if not (is_free(row + dr, col) and is_free(row, col + dc)):
                    continue
            out.append(((nr, nc), cost))
        return out

    def neighbors(self, s: Cell) -> list[tuple[Cell, float]]:
        # Ground-truth successors: enterable iff in bounds and actually free.
        return self._neighbors_impl(s, self.is_free_cell)

    # --- DynamicGridSpace[Cell] ------------------------------------------
    def passable_neighbors(self, s: Cell, blocked: set[Cell]) -> list[tuple[Cell, float]]:
        # Belief successors: enterable iff in bounds and not (yet) known blocked; real
        # occupancy is invisible here. Same worker + corner rule as neighbors().
        def believed_free(row: int, col: int) -> bool:
            return self.in_bounds(row, col) and (row, col) not in blocked

        return self._neighbors_impl(s, believed_free)

    def is_blocked(self, s: Cell) -> bool:
        # Occupied OR out of bounds — is_free_cell is already false for both.
        return not self.is_free_cell(*s)

    def heuristic(self, a: Cell, b: Cell) -> float:
        dr = abs(a[0] - b[0])
        dc = abs(a[1] - b[1])
        if self._connectivity == 8:
            # Octile distance: admissible for 8-connected unit/sqrt2 moves.
            # Evaluated as (hi - lo) + sqrt2*lo (identical operation order to the
            # C++ mirror) so f-values match bit-for-bit and tie-breaking, hence
            # trace event streams, stay identical across languages.
            lo = min(dr, dc)
            hi = max(dr, dc)
            return float(hi - lo) + _SQRT2 * float(lo)
        return float(dr + dc)  # Manhattan for 4-connected.

    # --- LineOfSightSpace[Cell] ------------------------------------------
    def line_of_sight(self, a: Cell, b: Cell) -> bool:
        # Straight segment between cell centres, tested by the same supercover
        # (Amanatides & Woo 1987) + corner-cut-forbidden rule as neighbors(), so a
        # LOS-visible pair is exactly a legal straight move (Nash et al. 2007).
        return self.is_motion_valid(self.cell_to_world(*a), self.cell_to_world(*b))

    # --- SamplingSpace[Point] --------------------------------------------
    def sample(self) -> Point:
        x = self._origin_x + float(self._rng.uniform(0.0, self._width * self._resolution))
        y = self._origin_y + float(self._rng.uniform(0.0, self._height * self._resolution))
        return (x, y)

    def is_state_valid(self, s: Point) -> bool:
        row, col = self.world_to_cell(s[0], s[1])
        return self.is_free_cell(row, col)

    # --- SE2CollisionSpace[Pose] -----------------------------------------
    def is_collision(self, footprint: Footprint, pose: Pose) -> bool:
        # Inscribed-disc footprint is orientation-invariant, so theta is unused (a
        # polygon footprint would use it) — Dolgov et al. 2008. Collision iff any
        # occupied or out-of-bounds cell overlaps the disc; exact disc–cell overlap
        # via squared distance to the cell rectangle (no sqrt, no trig → bit-identical
        # with the C++ mirror). world<->cell stays here.
        x, y, _theta = pose
        r = footprint.inscribed_radius
        r2 = r * r
        half = self._resolution * 0.5
        lo_row, lo_col = self.world_to_cell(x - r, y + r)  # y+r → smaller row
        hi_row, hi_col = self.world_to_cell(x + r, y - r)
        for row in range(lo_row, hi_row + 1):
            for col in range(lo_col, hi_col + 1):
                if self.is_free_cell(row, col):  # in-bounds & free → skip
                    continue
                cx, cy = self.cell_to_world(row, col)  # occupied OR out-of-bounds
                dx = x - min(max(x, cx - half), cx + half)
                dy = y - min(max(y, cy - half), cy + half)
                if dx * dx + dy * dy <= r2:
                    return True
        return False

    # --- ObstacleQuery (extends SE2CollisionSpace[Pose]) -------------------
    def _edt(self) -> np.ndarray:
        if self._edt_sq is None:
            h, w = self._height, self._width
            # 1-cell non-free border padding makes the map edge itself a source, so
            # a reactive planner near the boundary sees it as an obstacle (occupied
            # OR out-of-bounds already means "non-free" everywhere else in this class).
            padded = np.ones((h + 2, w + 2), dtype=bool)
            padded[1 : h + 1, 1 : w + 1] = ~self._free
            self._edt_sq = _squared_edt(padded)[1 : h + 1, 1 : w + 1]
        return self._edt_sq

    def distance_to_nearest(self, p: Point) -> float:
        row, col = self.world_to_cell(p[0], p[1])
        if not self.in_bounds(row, col):
            # p's own cell is itself non-free (out of bounds), so it is its own
            # nearest non-free cell.
            return 0.0
        # Squared distances are integer cell-unit values; sqrt is the only
        # floating-point step, so C++/Python agree up to that single rounding.
        return math.sqrt(self._edt()[row, col]) * self._resolution

    def occupied_within(self, center: Point, radius: float) -> list[Point]:
        # Same bounding-box convention as is_collision: world_to_cell(x-r, y+r) is
        # the lower row/col corner because +y maps to a smaller row index.
        x, y = center
        r2 = radius * radius
        lo_row, lo_col = self.world_to_cell(x - radius, y + radius)
        hi_row, hi_col = self.world_to_cell(x + radius, y - radius)
        out: list[Point] = []
        for row in range(lo_row, hi_row + 1):  # row asc, col asc: fixes summation
            for col in range(lo_col, hi_col + 1):  # order for PF/VFH force/bin sums
                if self.is_free_cell(row, col):  # in-bounds & free → not a source
                    continue
                cx, cy = self.cell_to_world(row, col)  # occupied OR out-of-bounds
                dx, dy = cx - x, cy - y
                if dx * dx + dy * dy <= r2:
                    out.append((cx, cy))
        return out

    def _is_free_uv(self, iu: int, iv: int) -> bool:
        # (u, v) grid coords count up from the origin (bottom-left); rows count
        # down from the top image row.
        return self.is_free_cell(self._height - 1 - iv, iu)

    def is_motion_valid(self, a: Point, b: Point) -> bool:
        # Supercover grid traversal (Amanatides & Woo, 1987): visits every cell
        # the segment crosses. Point sampling misses a corner clip whose in-cell
        # chord is shorter than the sample spacing, letting edges cut obstacle
        # corners.
        u0 = (a[0] - self._origin_x) / self._resolution
        v0 = (a[1] - self._origin_y) / self._resolution
        u1 = (b[0] - self._origin_x) / self._resolution
        v1 = (b[1] - self._origin_y) / self._resolution
        iu, iv = int(math.floor(u0)), int(math.floor(v0))
        ju, jv = int(math.floor(u1)), int(math.floor(v1))
        if not self._is_free_uv(iu, iv):
            return False
        du, dv = u1 - u0, v1 - v0
        step_u = 1 if du > 0.0 else -1
        step_v = 1 if dv > 0.0 else -1
        t_delta_u = abs(1.0 / du) if du != 0.0 else math.inf
        t_delta_v = abs(1.0 / dv) if dv != 0.0 else math.inf
        # Parametric distance (in units of t) from the start to the first grid
        # line crossed on each axis; an axis with no motion never crosses one
        # (explicit inf also avoids 0 * inf = nan when the start sits on a line).
        if du != 0.0:
            frac_u = (math.floor(u0) + 1.0 - u0) if du > 0.0 else (u0 - math.floor(u0))
            t_max_u = frac_u * t_delta_u
        else:
            t_max_u = math.inf
        if dv != 0.0:
            frac_v = (math.floor(v0) + 1.0 - v0) if dv > 0.0 else (v0 - math.floor(v0))
            t_max_v = frac_v * t_delta_v
        else:
            t_max_v = math.inf
        while iu != ju or iv != jv:
            # Clamp exhausted axes so float drift in t_max can never step past
            # the end cell (termination is by cell index, not by t).
            if iv == jv or t_max_u < t_max_v:
                iu += step_u
                t_max_u += t_delta_u
            elif iu == ju or t_max_v < t_max_u:
                iv += step_v
                t_max_v += t_delta_v
            else:
                # Exact corner crossing: same rule as neighbors() — passing
                # through a corner needs both shared orthogonal cells free.
                if not (self._is_free_uv(iu + step_u, iv) and self._is_free_uv(iu, iv + step_v)):
                    return False
                iu += step_u
                iv += step_v
                t_max_u += t_delta_u
                t_max_v += t_delta_v
            if not self._is_free_uv(iu, iv):
                return False
        return True

    def distance(self, a: Point, b: Point) -> float:
        return math.hypot(b[0] - a[0], b[1] - a[1])

    def steer(self, a: Point, b: Point, eta: float) -> Point:
        dist = self.distance(a, b)
        if dist <= eta or dist == 0.0:
            return b
        scale = eta / dist
        return (a[0] + (b[0] - a[0]) * scale, a[1] + (b[1] - a[1]) * scale)

File: cpp/src/maps/occupancy_grid.cpp
#include "navigation/maps/occupancy_grid.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>

namespace navigation::maps {
namespace {

// Sentinel for "no source seen yet" in the 1D transform below. Squared cell
// distances on any map we load are tiny by comparison, so this never collides
// with a real value.
constexpr double kEdtInf = 1e18;

// Exact 1D squared-distance transform, in place: f[i] = 0 marks a source, any
// other value is replaced by the squared distance (in index units) to the
// nearest source. Lower-envelope-of-parabolas algorithm (Felzenszwalb &
// Huttenlocher 2004, "Distance Transforms of Sampled Functions") — O(n), exact
// (integer-valued results, only the caller's final sqrt is inexact).
void edt_1d(std::vector<double>& f) {
  const int n = static_cast<int>(f.size());
  std::vector<int> v(n, 0);
  std::vector<double> z(static_cast<size_t>(n) + 1);
  std::vector<double> d(n);
  int k = 0;
  v[0] = 0;
  z[0] = -kEdtInf;
  z[1] = kEdtInf;
  for (int q = 1; q < n; ++q) {
    double s = ((f[q] + static_cast<double>(q) * q) -
                (f[v[k]] + static_cast<double>(v[k]) * v[k])) /
               (2.0 * (q - v[k]));
    while (s <= z[k]) {
      --k;
      s = ((f[q] + static_cast<double>(q) * q) -
           (f[v[k]] + static_cast<double>(v[k]) * v[k])) /
          (2.0 * (q - v[k]));
    }
    ++k;
    v[k] = q;
    z[k] = s;
    z[k + 1] = kEdtInf;
  }
  k = 0;
  for (int q = 0; q < n; ++q) {
    while (z[k + 1] < q) ++k;
    double diff = q - v[static_cast<size_t>(k)];
    d[q] = diff * diff + f[static_cast<size_t>(v[k])];
  }
  f = std::move(d);
}

// Squared-distance-to-nearest-source grid, in cell units, over a (rows+2) x
// (cols+2) padded grid: the outer 1-cell ring is entirely source (models "out
// of bounds" as immediately adjacent, matching is_collision's occupied-OR-out-
// of-bounds rule); interior cells are source iff occupied. Row-major, size
// (rows+2)*(cols+2). Two 1D passes (columns then rows) give the exact 2D
// Euclidean squared-distance transform.
std::vector<double> compute_edt(int rows, int cols, const std::vector<bool>& free) {
  const int pr = rows + 2, pc = cols + 2;
  std::vector<double> g(static_cast<size_t>(pr) * pc);
  for (int i = 0; i < pr; ++i) {
    for (int j = 0; j < pc; ++j) {
      bool source = i == 0 || i == pr - 1 || j == 0 || j == pc - 1 ||
                    !free[static_cast<size_t>(i - 1) * cols + (j - 1)];
      g[static_cast<size_t>(i) * pc + j] = source ? 0.0 : kEdtInf;
    }
  }
  std::vector<double> col(pr);
  for (int j = 0; j < pc; ++j) {
    for (int i = 0; i < pr; ++i) col[i] = g[static_cast<size_t>(i) * pc + j];
    edt_1d(col);
    for (int i = 0; i < pr; ++i) g[static_cast<size_t>(i) * pc + j] = col[i];
  }
  std::vector<double> row(pc);
  for (int i = 0; i < pr; ++i) {
    for (int j = 0; j < pc; ++j) row[j] = g[static_cast<size_t>(i) * pc + j];
    edt_1d(row);
    for (int j = 0; j < pc; ++j) g[static_cast<size_t>(i) * pc + j] = row[j];
  }
  return g;
}

}  // namespace

OccupancyGrid2D::OccupancyGrid2D(int rows, int cols, double resolution, double origin_x,
                                 double origin_y, std::vector<bool> free_cells, int connectivity,
                                 unsigned seed)
    : rows_(rows),
      cols_(cols),
      resolution_(resolution),
      origin_x_(origin_x),
      origin_y_(origin_y),
      free_(std::move(free_cells)),
      connectivity_(connectivity),
      rng_(seed) {
  if (rows_ <= 0 || cols_ <= 0) throw std::runtime_error("grid: non-positive size");
  if (free_.size() != static_cast<size_t>(rows_) * static_cast<size_t>(cols_)) {
    throw std::runtime_error("grid: free-cell count mismatch");
  }
  if (connectivity_ != 4 && connectivity_ != 8) {
    throw std::runtime_error("grid: connectivity must be 4 or 8");
  }
}

OccupancyGrid2D OccupancyGrid2D::from_image(const PgmImage& img, double resolution, double origin_x,
                                            double origin_y, double occupied_thresh,
                                            double free_thresh, int connectivity, unsigned seed) {
  (void)occupied_thresh;  // free_thresh alone decides traversability; occupied/unknown are blocked.
  std::vector<bool> free_cells(img.pixels.size());
  for (size_t i = 0; i < img.pixels.size(); ++i) {
    double occ = 1.0 - static_cast<double>(img.pixels[i]) / 255.0;
    free_cells[i] = occ <= free_thresh;
  }
  return OccupancyGrid2D(img.height, img.width, resolution, origin_x, origin_y,
                         std::move(free_cells), connectivity, seed);
}

std::set<Capability> OccupancyGrid2D::capabilities() const {
  return {Capability::DISCRETE_SPACE,      Capability::SAMPLING_SPACE,
          Capability::LINE_OF_SIGHT_SPACE, Capability::DYNAMIC_GRID_SPACE,
          Capability::SE2_COLLISION_SPACE, Capability::OBSTACLE_QUERY};
}

bool OccupancyGrid2D::in_bounds(int row, int col) const {
  return row >= 0 && row < rows_ && col >= 0 && col < cols_;
}

bool OccupancyGrid2D::is_free(int row, int col) const {
  return in_bounds(row, col) && free_[static_cast<size_t>(row) * cols_ + col];
}

Point OccupancyGrid2D::cell_to_world(const Cell& c) const {
  double x = origin_x_ + (c.col + 0.5) * resolution_;
  double y = origin_y_ + ((rows_ - 1 - c.row) + 0.5) * resolution_;
  return {x, y};
}

Cell OccupancyGrid2D::world_to_cell(double x, double y) const {
  int col = static_cast<int>(std::floor((x - origin_x_) / resolution_));
  int row = (rows_ - 1) - static_cast<int>(std::floor((y - origin_y_) / resolution_));
  return {row, col};
}

std::vector<std::pair<Cell, double>> OccupancyGrid2D::neighbors(const Cell& s) const {
  // Ground-truth successors: a cell is enterable iff in bounds and actually free.
  return neighbors_impl(s, [this](int row, int col) { return is_free(row, col); });
}

std::vector<std::pair<Cell, double>> OccupancyGrid2D::passable_neighbors(
    const Cell& s, const std::set<Cell>& blocked) const {
  // Belief successors: a cell is enterable iff in bounds and not (yet) known blocked;
  // real occupancy is invisible here (only is_blocked reads it). Same worker + corner
  // rule as neighbors(), so a discovered-free grid gives identical successors to truth.
  return neighbors_impl(s, [this, &blocked](int row, int col) {
    return in_bounds(row, col) && blocked.find(Cell{row, col}) == blocked.end();
  });
}

bool OccupancyGrid2D::is_blocked(const Cell& s) const {
  // Occupied OR out of bounds — is_free is already false for both.
  return !is_free(s.row, s.col);
}

double OccupancyGrid2D::heuristic(const Cell& a, const Cell& b) const {
  int dr = std::abs(a.row - b.row);
  int dc = std::abs(a.col - b.col);
  if (connectivity_ == 4) return dr + dc;  // Manhattan is admissible for 4-connected moves.
  // Octile distance: admissible lower bound for 8-connected moves with diagonal
  // cost sqrt(2) (D = 1, D2 = sqrt(2)).
  int lo = std::min(dr, dc);
  int hi = std::max(dr, dc);
  return (hi - lo) + std::sqrt(2.0) * lo;
}

bool OccupancyGrid2D::line_of_sight(const Cell& a, const Cell& b) const {
  // Any-angle LOS must mean "the straight segment is actually traversable" under
  // the SAME corner-cut-forbidden rule as neighbors()/is_motion_valid — else a
  // shortcut could clip an obstacle corner the grid edges forbid. Reuse the
  // verified supercover (Amanatides & Woo 1987) from cell centres; world
  // conversion stays inside the map (Nash, Daniel, Koenig & Felner 2007).
  return is_motion_valid(cell_to_world(a), cell_to_world(b));
}

Point OccupancyGrid2D::sample() {
  std::uniform_real_distribution<double> dx(origin_x_, origin_x_ + cols_ * resolution_);
  std::uniform_real_distribution<double> dy(origin_y_, origin_y_ + rows_ * resolution_);
  return {dx(rng_), dy(rng_)};
}

bool OccupancyGrid2D::is_state_valid(const Point& p) const {
  Cell c = world_to_cell(p.x, p.y);
  return is_free(c.row, c.col);
}

bool OccupancyGrid2D::is_collision(const Footprint& fp, const Pose& pose) const {
  // Inscribed-disc footprint is orientation-invariant, so pose.theta is unused (a
  // polygon footprint would use it) — Dolgov et al. 2008. Collision iff any occupied
  // or out-of-bounds cell overlaps the disc of radius r at (x, y). Exact disc–cell
  // overlap via squared distance to the cell rectangle (no sqrt, no trig → bit-
  // identical across languages). Fixed row-major scan; world<->cell stays here.
  const double r = fp.inscribed_radius, r2 = r * r, half = resolution_ * 0.5;
  const Cell lo = world_to_cell(pose.x - r, pose.y + r);  // y+r → smaller row
  const Cell hi = world_to_cell(pose.x + r, pose.y - r);
  for (int row = lo.row; row <= hi.row; ++row) {
    for (int col = lo.col; col <= hi.col; ++col) {
      if (is_free(row, col)) continue;               // in-bounds & free → skip
      const Point c = cell_to_world(Cell{row, col});  // occupied OR out-of-bounds
      const double dx = pose.x - std::clamp(pose.x, c.x - half, c.x + half);
      const double dy = pose.y - std::clamp(pose.y, c.y - half, c.y + half);
      if (dx * dx + dy * dy <= r2) return true;
    }
  }
  return false;
}

double OccupancyGrid2D::distance_to_nearest(const Point& p) const {
  if (!edt_) edt_ = compute_edt(rows_, cols_, free_);
  const Cell c = world_to_cell(p.x, p.y);
  // Padded index = grid index + 1 (the added border ring). is_free() already
  // treats every out-of-bounds cell as non-free regardless of how far out it
  // is, so a cell beyond the ring is itself a source — clamping into the ring
  // yields exactly that (distance 0), while also keeping the lookup in bounds.
  const int pr = std::clamp(c.row + 1, 0, rows_ + 1);
  const int pc = std::clamp(c.col + 1, 0, cols_ + 1);
  const double d2 = (*edt_)[static_cast<size_t>(pr) * (cols_ + 2) + pc];
  return std::sqrt(d2) * resolution_;
}

std::vector<Point> OccupancyGrid2D::occupied_within(const Point& center, double radius) const {
  // Same bounding-box-then-filter shape as is_collision: convert the world
  // radius to a cell-index box via world_to_cell, then keep non-free cells
  // whose center actually falls within radius. Row ascending, then column
  // ascending (outer/inner loop order) — enumeration order is part of the
  // capability contract (core/capabilities.hpp), not an implementation detail.
  const double r2 = radius * radius;
  const Cell lo = world_to_cell(center.x - radius, center.y + radius);  // y+r -> smaller row
  const Cell hi = world_to_cell(center.x + radius, center.y - radius);
  std::vector<Point> out;
  for (int row = lo.row; row <= hi.row; ++row) {
    for (int col = lo.col; col <= hi.col; ++col) {
      if (is_free(row, col)) continue;                 // in-bounds & free -> skip
      const Point c = cell_to_world(Cell{row, col});    // occupied OR out-of-bounds
      const double dx = c.x - center.x, dy = c.y - center.y;
      if (dx * dx + dy * dy <= r2) out.push_back(c);
    }
  }
  return out;
}

bool OccupancyGrid2D::is_free_uv(int iu, int iv) const {
  // (u, v) grid coords count up from the origin (bottom-left); rows count down
  // from the top image row.
  return is_free(rows_ - 1 - iv, iu);
}

bool OccupancyGrid2D::is_motion_valid(const Point& a, const Point& b) const {
  // Supercover grid traversal (Amanatides & Woo, 1987): visits every cell the
  // segment crosses. Point sampling misses a corner clip whose in-cell chord is
  // shorter than the sample spacing, letting edges cut obstacle corners.
  const double kInf = std::numeric_limits<double>::infinity();
  double u0 = (a.x - origin_x_) / resolution_, v0 = (a.y - origin_y_) / resolution_;
  double u1 = (b.x - origin_x_) / resolution_, v1 = (b.y - origin_y_) / resolution_;
  int iu = static_cast<int>(std::floor(u0)), iv = static_cast<int>(std::floor(v0));
  int ju = static_cast<int>(std::floor(u1)), jv = static_cast<int>(std::floor(v1));
  if (!is_free_uv(iu, iv)) return false;
  double du = u1 - u0, dv = v1 - v0;
  int step_u = du > 0.0 ? 1 : -1;
  int step_v = dv > 0.0 ? 1 : -1;
  double t_delta_u = du != 0.0 ? std::abs(1.0 / du) : kInf;
  double t_delta_v = dv != 0.0 ? std::abs(1.0 / dv) : kInf;
  // Parametric distance (in units of t) from the start to the first grid line
  // crossed on each axis; an axis with no motion never crosses one (explicit
  // inf also avoids 0 * inf = nan when the start sits on a line).
  double t_max_u =
      du != 0.0 ? (du > 0.0 ? std::floor(u0) + 1.0 - u0 : u0 - std::floor(u0)) * t_delta_u : kInf;
  double t_max_v =
      dv != 0.0 ? (dv > 0.0 ? std::floor(v0) + 1.0 - v0 : v0 - std::floor(v0)) * t_delta_v : kInf;
  while (iu != ju || iv != jv) {
    // Clamp exhausted axes so float drift in t_max can never step past the end
    // cell (termination is by cell index, not by t).
    if (iv == jv || t_max_u < t_max_v) {
      iu += step_u;
      t_max_u += t_delta_u;
    } else if (iu == ju || t_max_v < t_max_u) {
      iv += step_v;
      t_max_v += t_delta_v;
    } else {
      // Exact corner crossing: same rule as neighbors() — passing through a
      // corner needs both shared orthogonal cells free.
      if (!is_free_uv(iu + step_u, iv) || !is_free_uv(iu, iv + step_v)) return false;
      iu += step_u;
      iv += step_v;
      t_max_u += t_delta_u;
      t_max_v += t_delta_v;
    }
    if (!is_free_uv(iu, iv)) return false;
  }
  return true;
}

double OccupancyGrid2D::distance(const Point& a, const Point& b) const {
  return std::hypot(a.x - b.x, a.y - b.y);
}

Point OccupancyGrid2D::steer(const Point& a, const Point& b, double eta) const {
  double d = distance(a, b);
  if (d <= eta || d == 0.0) return b;
  double s = eta / d;
  return {a.x + s * (b.x - a.x), a.y + s * (b.y - a.y)};
}

}  // namespace navigation::maps
